added column sorting feature on table

tyrone-wu opened the pull request:
Added functionality for sorting table based on a column. #2 

Most of the added code is for the new sorting feature. 
I modified the `BpfProgram` class with additional fields from the columns (period avg runtime, total avg runtime, events per sec, cpu time percent) so that they don't need to be computed twice when first displaying to UI, and then when sorting. 
I also added a `new` method for initializing a `BpfProgram` object so that it doesn't take up too much lines (which was in conjunction to the additional fields).

tyrone-wu commented:
I don't have as much programs to work with at the moment, but here's how it looks like:
![filters](https://github.com/Netflix/bpftop/assets/33705740/0439e1bc-570e-4c38-8e74-bcc9cc337caf)

jfernandez commented:
@tyrone-wu I think you pasted the gif for the filter feature.

I tried the feature locally and it's working great. The only thing I'd add functionally is to make Enter do the following when in sorting mode: 
* If no sorting is enabled on the column, sort by DESC and exit sorting mode
* If sorting is already enabled, swap the direction and exit sorting mode

I don't think we need to add this to the footer, just make it a hidden feature for people that may want to do this due to muscle memory from other top-like apps.

tyrone-wu commented:
The new hidden feature is now added! 👍 
I also corrected the gif from earlier 😅, and here's the hidden feature in action. 
![sort_hidden_feature](https://github.com/Netflix/bpftop/assets/33705740/dce2ea14-bdd7-4993-9250-0ece9f62d2e5)

jfernandez commented on src/app.rs:
```diff
@@ -29,15 +29,20 @@ use tui_input::Input;
 
 use crate::bpf_program::BpfProgram;
 
+const NUM_COLUMNS: i8 = 7;
```
I see that we used to have a HEADER_COLUMNS constant in main.rs and then dropped in this PR. Can we bring that back and then ask its size? I worry that by having more than one place where we store the number of columns may lead to bugs.

tyrone-wu replied:
The main.rs `HEADER_COLUMNS` was moved into app.rs as `header_columns` field. This was so that the UI doesn't have to re-concat the header text and sort symbol in the draw loop. 
In the PR, app.rs now handles the text headers, which also handles applying the appropriate sort symbol to the selected header column. The main.rs now just renders the header text from the `header_columns` field in app.rs. 

But I updated it so that the `i8` const is no longer needed. It should now only rely on `header_columns.len()` when getting the size.

jfernandez commented on src/bpf_program.rs:
```diff
@@ -70,12 +99,19 @@ impl BpfProgram {
         self.runtime_delta() as f64 / self.period_ns as f64 * 100.0
     }
+
+    pub fn update_stats(&mut self) {
```
Can we sort without having to cache the value? I would prefer not doing this caching unless it was absolutely necessary

tyrone-wu replied:
The sort is now updated without using the caching, and bpf_program.rs should now be reverted to what it previously was as well. 👍

jfernandez commented on src/main.rs:
```diff
@@ -432,44 +460,64 @@ fn render_table(f: &mut Frame, app: &mut App, area: Rect) {
-                filter_area[0].x + filter_input.visual_cursor() as u16 + 2,
-                filter_area[0].y + 1,
-            )
+    if let Mode::Table | Mode::Graph = app.mode {
```
This line confuses me, what exactly is it doing?

tyrone-wu replied:
This is another way to doing enum pattern matching, similar to `if app.mode == Mode::Table || app.mode == Mode::Graph`, or:
```rust
match app.mode {
    Mode::Table | Mode::Graph => ...,
```

I made a slight modification to this part so that it's less "nested", if it helps improve the readability.

jfernandez reviewed (changes requested):
Functionally this is working great. I had some questions about some design choices.
